src/material_library.cpp

```diff
@@ -0,0 +1,391 @@
+#include <unistd.h>
+#include <iostream>
+
+#ifndef PYNE_IS_AMALGAMATED
+#include "material_library.h"
+#endif
+
+// Empty Constructor
+pyne::MaterialLibrary::MaterialLibrary(){};
+
+// Default constructor
+pyne::MaterialLibrary::MaterialLibrary(const std::string& file,
+                                       const std::string& datapath) {
+  if (!pyne::file_exists(file)) {
+    throw std::runtime_error("File " + file +
+                             " not found or no read permission");
+  }
+  if (!hdf5_path_exists(file, datapath)) {
+    throw std::runtime_error("The datapath, " + datapath + ", in " + file +
+                             " is empty.");
+  }
+  // load materials
+  from_hdf5(file, datapath);
+};
+
+void pyne::MaterialLibrary::from_hdf5(char* fname, char* dpath, char* npath,
+                                      int protocol) {
+  std::string filename(fname);
+  std::string datapath(dpath);
+  std::string nucpath(npath);
+  from_hdf5(filename, datapath, npath, protocol);
+}
+// Append Material to the library from hdf5 file
+void pyne::MaterialLibrary::from_hdf5(const std::string& filename,
+                                      const std::string& datapath,
+                                      const std::string& nucpath,
+                                      int protocol) {
+  if (!hdf5_path_exists(filename, datapath)) return;
```

Comment:
Suggestion: print a warning message to tell the users that they are trying to append material from an empty library.